Mathlib/CategoryTheory/Adjunction/Basic.lean

```diff
@@ -176,6 +176,41 @@ theorem homEquiv_naturality_right_symm (f : X ⟶ G.obj Y) (g : Y ⟶ Y') :
   simp only [homEquiv_naturality_right,eq_self_iff_true,Equiv.apply_symm_apply]
 #align category_theory.adjunction.hom_equiv_naturality_right_symm CategoryTheory.Adjunction.homEquiv_naturality_right_symm
 
+@[reassoc]
+theorem homEquiv_naturality_left_square (f : X' ⟶ X) (g : F.obj X ⟶ Y')
+    (h : F.obj X' ⟶ Y) (k : Y ⟶ Y') (w : F.map f ≫ g = h ≫ k) :
+    f ≫ (adj.homEquiv X Y') g = (adj.homEquiv X' Y) h ≫ G.map k := by
+  rw [← homEquiv_naturality_left, ← homEquiv_naturality_right, w]
+
+@[reassoc]
+theorem homEquiv_naturality_right_square (f : X' ⟶ X) (g : X ⟶ G.obj Y')
+    (h : X' ⟶ G.obj Y) (k : Y ⟶ Y') (w : f ≫ g = h ≫ G.map k) :
+    F.map f ≫ (adj.homEquiv X Y').symm g = (adj.homEquiv X' Y).symm h ≫ k := by
+  rw [← homEquiv_naturality_left_symm, ← homEquiv_naturality_right_symm, w]
+
+theorem homEquiv_naturality_left_square_iff (f : X' ⟶ X) (g : F.obj X ⟶ Y')
+    (h : F.obj X' ⟶ Y) (k : Y ⟶ Y') :
+    (F.map f ≫ g = h ≫ k) ↔
+    (f ≫ (adj.homEquiv X Y') g = (adj.homEquiv X' Y) h ≫ G.map k) where
+  mp := homEquiv_naturality_left_square adj f g h k
+  mpr := by
+    intro w
+    have transposedw := homEquiv_naturality_right_square adj _ _ _ _ w
+    simp only [Equiv.symm_apply_apply] at transposedw
+    exact transposedw
```

Comment:
```suggestion
    (f ≫ (adj.homEquiv X Y') g = (adj.homEquiv X' Y) h ≫ G.map k) ↔
      (F.map f ≫ g = h ≫ k) :=
  ⟨fun w ↦ by simpa only [Equiv.symm_apply_apply]
      using homEquiv_naturality_right_square adj _ _ _ _ w,
    homEquiv_naturality_left_square adj f g h k⟩
```

I suggest stating the lemma in the other direction (but unfortunately it cannot be made a `simp` lemma).

There is nothing wrong with the `where mp := ... mpr := ...` syntax, but in mathlib, we are more used to term proofs as in my suggestion when the proofs are very short, or, by using the following syntax in tactic mode otherwise:
```lean
  constructor
  · sorry -- first implication
  · sorry -- second implication
```

About, the `have transposedw := ...` syntax, when there is only one such `have`, we may use `have := ...` and use `this` instead of an explicit name. In my suggestion, I have used the tactic `simpa ... using ...`, which basically apply simp lemmas to given term until it hopefully matches the goal.